rose/server/score.py

```diff
@@ -19,62 +17,72 @@ def process(players, track):
-        player.score += config.score_move_forward
-        obstacle = track.get(player.x, player.y)
-        if obstacle == obstacles.CRACK:
-            if player.action != actions.JUMP:
-                track.clear(player.x, player.y)
-                player.y += 1
-                player.score += config.score_move_backward * 2
-            else:
-                player.score += config.score_jump
         elif obstacle in (obstacles.TRASH,
                           obstacles.BIKE,
                           obstacles.BARRIER):
-            if player.action not in (actions.LEFT, actions.RIGHT):
+            # Move back
+            track.clear(player.x, player.y)
+            player.y += 1
+            player.score += config.score_move_backward
+
+        elif obstacle == obstacles.CRACK:
+            if player.action == actions.JUMP:
+                # Move forlward
+                player.score += config.score_move_forward
+                player.score += config.score_jump
+            else:
+                # Move back
                 track.clear(player.x, player.y)
                 player.y += 1
-                player.score += config.score_move_backward * 2
+                player.score += config.score_move_backward
+
         elif obstacle == obstacles.WATER:
-            if player.action != actions.BRAKE:
+            if player.action == actions.BRAKE:
+                # Move forlward
+                player.score += config.score_move_forward
+                player.score += config.score_brake
+            else:
+                # Move back
                 track.clear(player.x, player.y)
                 player.y += 1
-                player.score += config.score_move_backward * 2
-            else:
-                player.score += config.score_brake
+                player.score += config.score_move_backward
+
         elif obstacle == obstacles.PENGUIN:
             if player.action == actions.PICKUP:
+                # Move forlward
                 track.clear(player.x, player.y)
                 player.score += config.score_move_forward
+                player.score += config.score_pickup
+            else:
+                # Move forlward no reward
+                player.score += config.score_move_forward
 
-        # Here we can end the game when player gets out of
-        # the track bounds. For now, just keep the player at the same
-        # location.
-        player.y = min(config.matrix_height - 1, max(2, player.y))
-
-        # Finally forget action
-        player.action = actions.NONE
+        # Filter for players with the same x and y
+        all_collisions = filter(lambda p:
+                                p.name != player.name and
+                                p.x == player.x and
+                                p.y == player.y, players.values())
+        next_collision = next(all_collisions, None)
 
-        # Fix up collisions
+        # if we have a collision and we are not in our lane, try to move back
```

Comment:
This depends on order of players, this is why we process players in some order. The previous player got this position so the next players need to find another position.

Reply:
The documentation say:

>NOTE : If you are driving out of your lane and have a collision with the lane owner, you will punished by -10 points and be moved to other free place.

so this is the currect behavior according to the documentation

Ref:
https://github.com/RedHat-Israel/ROSE/tree/master/examples

Reply:
The documentation does not describe exactly the semantics. Handling collisions is much more complicated than what we describe in the docs.

We process the gate state in 3 steps:
1. Change the location, needed only for RIGHT or LEFT actions. We can see this as our physics engine :-). This is done in no particular order.
2. Handle obstacles - this must be done in order, since when a player pick up or hit an obstacle, we remove the obstacle from the track. The next player will not hit this obstacle since it does not exist at this point.
3. Fix up collisions by moving the player to another position if another player got this position before.

You can see this more clearly using the new debug logs I added in #479.

We can change the way this works, but we can also change the documentation to describe better how collision are handled.

You propose many change in this pr:
> Use positive logic to follow the documentation, e.g. replace the != with == and not in with in

This sounds like good change for entire pr.

> When testing collisions xplicitly test for in_lane instead of using the implicit sort technique

I'm not sure what implicit about the current code, and I don't see how the suggested code is better and how it handles correctly obstacles, so only the first player pick/hit the obstacle.

> Use score_pickup explicitly, instead of using score_move_forward

Sounds good, simple change that should be in its own commit.

> When moving backword use ``score_move_backward instead of adding score_move_forward
and then reducing `score_move_backward` twice.

Sounds good for its own commit - the current code is confusing. It is better if we change the score in single place at the processing loop.

> Move action clean up to the end of the method

Sounds good for its own commit. Maybe this should happen at the point we process the action - once we processed the action it can be set to None so we cannot process the same action twice.

I suggest to focus on the simple changes that cannot break anything first to get rid of them and make the code cleaner before we try to handle the bigger changes, where we are not sure if we need to change the code or the docs.

Reply:
> I suggest to focus on the simple changes that cannot break anything first to get rid of them and make the code cleaner before we try to handle the bigger changes, where we are not sure if we need to change the code or the docs.

ok, reverted back to using the sorting method, and added a comment explaining it's logic